network/tls.go

```diff
@@ -0,0 +1,415 @@
+package network
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"math/big"
+	"net"
+	"time"
+
+	"github.com/dedis/kyber/sign/schnorr"
+	"github.com/dedis/kyber/util/encoding"
+	"github.com/dedis/kyber/util/random"
+	"github.com/dedis/onet/log"
+)
+
+// About our TLS strategy:
+//
+// The design goals were:
+// 1. use vanilla TLS for conode/conode communication, in order to ease deployment
+//    anxiety with industrial partners. ("If it's not TLS, it's not allowed." Sigh.)
+// 2. zero config: whatever config we have in an existing private.toml should be enough
+// 3. mutual authentication: each side should end up with proof that the other side is
+//    holding the secret key associated with the public key they claim to have.
+//
+// In order to achieve #1, we limit ourselves to TLS 1.2. This means that the TLS
+// private key must be different than the conode private key (which may be from a suite
+// not supported in TLS 1.2).
+//
+// In order to achieve #2, we use self-signed TLS certificates, with a private key
+// that is created on server boot and stored in RAM. Because the certificates are
+// self-signed, there is no need to pay a CA, nor to load the private key
+// or certificate from disk.
+//
+// In order to achieve #3, we include an extension in the certificate, which
+// proves that the same entity that holds the TLS private key (i.e. the signing
+// key for the self-signed TLS cert) also holds the conode's private key.
+// We do this by hashing a nonce provided by the peer (in order to prove that this
+// is a fresh challenge response) and the ASN.1 encoded CommonName of the certificate.
+// The CN is always the hex-encoded form of the conode's public key.
+//
+// Because each side needs a nonce which is controlled by the opposite party in the
+// mutual authentication, but TLS does not support sending application data before the handshake,
+// we need to find places in the normal TLS 1.2 handshake where we can "tunnel" the nonce
+// through. On the TLS client side, we use ClientHelloInfo.ServerName. On the
+// the TLS server side, we use the.
+//
+// There is a risk that with a less customizable TLS implementation than
+// Go's, it would not be possible to send the nonces through like this. However,
+// for the moment we are not targeting other languages than Go on the conode/conode
+// communication channel.
+
+// TODO: Websockets.
+// All of this is completely unrelated to HTTPS security on the websocket side. For
+// that, we will implement an opt-in Let's Encrypt client in websocket.go.
+
+// certMaker holds the data necessary to make a certificate on the fly
+// and give it to crypto/tls via the GetCertificate and
+// GetClientCertificate callbacks.
```

Comment:
`GetCertificate` and `GetClientCertificate` are private now

Reply:
I was talking about the callbacks in the tls.Config structure. I fixed the comment to be more precise.